src/main/scala/de/unruh/isabelle/pure/Position.scala

```diff
@@ -24,10 +40,45 @@ object Position extends MLValueWrapper.Companion[Position] {
 
   protected class Ops(implicit isabelle: Isabelle) extends super.Ops {
     lazy val none: Position = compileValue[Position]("Position.none").retrieveNow
+    lazy val start: Position = compileValue[Position]("Position.start").retrieveNow
+    lazy val startWithFileName = compileFunction[String, Position]("Position.file")
+    lazy val startWithId = compileFunction[String, Position]("Position.id")
+
+    lazy val lineOf = compileFunction[Position, Option[Int]]("Position.line_of")
+    lazy val offsetOf = compileFunction[Position, Option[Int]]("Position.offset_of")
+    lazy val endOffsetOf = compileFunction[Position, Option[Int]]("Position.end_offset_of")
+    lazy val fileOf = compileFunction[Position, Option[String]]("Position.file_of")
+    lazy val idOf = compileFunction[Position, Option[String]]("Position.id_of")
+
+    lazy val extract = compileFunction[Position, String, String](
+      """fn (pos, s) =>
+        | case (Position.offset_of pos, Position.end_offset_of pos) of
+        |   (SOME istart, SOME iend) => Symbol.explode s |> Library.drop (istart - 1) |> Library.take (iend - istart) |> String.concat
+        | | (SOME istart, NONE) => Symbol.explode s |> Library.drop (istart - 1) |> String.concat
+        | | _ => raise Fail "Position.extract: no offset"
+        |""".stripMargin
+    )
+    lazy val extractRange = compileFunction[Position, Position, String, String](
+      """fn (start_pos, end_pos, s) =>
+        | case (Position.offset_of start_pos, Position.offset_of end_pos) of
+        |   (SOME istart, SOME iend) => Symbol.explode s |> Library.drop (istart - 1) |> Library.take (iend - istart) |> String.concat
+        | | (SOME istart, NONE) => Symbol.explode s |> Library.drop (istart - 1) |> String.concat
+        | | _ => raise Fail "Position.extract: no offset"
+        |""".stripMargin
+    )
   }
 
   /** Represents an unspecified position (`Position.none` in ML). */
   def none(implicit isabelle: Isabelle): Position = Ops.none
+  /** Represents a starting position with no associated filename or ID (`Position.start` in ML). */
+  def start(implicit isabelle: Isabelle): Position = Ops.start
+  /** Represents the starting position in a given file (`Position.file` in ML). */
+  def startWithFileName(fileName: String)(implicit isabelle: Isabelle): Position = Ops.startWithFileName(fileName).retrieveNow
+  /** Represents the starting position in a given text (`Position.id` in ML). */
+  def startWithId(id: String)(implicit isabelle: Isabelle): Position = Ops.startWithId(id).retrieveNow
+
+  /** Returns the substring from this position's offset (inclusive) to another's offset (exclusive). */
+  def extractRange(start: Position, end: Position, text: String)(implicit isabelle: Isabelle): String = Ops.extractRange(start, end, text).retrieveNow
```

Comment:
Maybe move this into `class Position` (using `this` instead of `start`) and rename it to `extractUntil`? So a call would be `start.extractUntil(end, text)`.

Reply:
Ok